lib/task-graph.js

```diff
@@ -0,0 +1,547 @@
+                });
+            } else if (taskData.taskDefinition) {
+                var definition = self.constructInlineTaskObject(taskData.taskDefinition,
+                    taskOverrides, taskData.optionOverrides, taskData.label);
+                self.tasks[definition.instanceId] = definition;
+            }
+        })
+        .spread(function() {
+            return self;
+        });
+    };
+
+    TaskGraph.prototype.constructInlineTaskObject = function(_definition, taskOverrides,
+           optionOverrides, label) {
+
+        return this._buildTaskDefinition(_definition, optionOverrides,
+                taskOverrides, label);
+    };
+
+    TaskGraph.prototype.constructTaskObject = function(taskName, taskOverrides,
+           optionOverrides, label) {
+        var self = this;
+        return store.getTaskDefinition(taskName)
+        .then(function(taskDefinition) {
+            return self._buildTaskDefinition(
+                taskDefinition,
+                optionOverrides,
+                taskOverrides,
+                label
+            );
+        });
+    };
+
+    TaskGraph.prototype._buildTaskDefinition = function(_definition, optionOverrides,
+            taskOverrides, label) {
+        var self = this;
+        var definition = _.cloneDeep(_definition);
+
+        var baseTaskDefinition = self._getBaseTask(_definition);
+        definition.instanceId = taskOverrides.instanceId;
+        definition.properties = _.merge(definition.properties, baseTaskDefinition.properties);
+        definition.runJob = baseTaskDefinition.runJob;
+        definition.options = _.merge(definition.options || {}, optionOverrides || {});
+        definition.label = label;
+        definition.name = taskOverrides.name || definition.injectableName;
+        definition.waitingOn = taskOverrides.waitingOn || {};
+        // TODO: Remove ignoreFailure in favor of better graph branching evalution.
+        // NOTE: actually ignoreFailure is still useful for tasks that can
+        // fail but we don't care AND don't have anything queued up to run
+        // on failure (rare case probably, but still worth supporting IMO).
+        definition.ignoreFailure = taskOverrides.ignoreFailure || false;
+
+        var allOptions = _.uniq(
+            _.keys(definition.options).concat(baseTaskDefinition.requiredOptions)
+        );
+
+        // If the graph has specifically defined options for a task, don't bother
+        // with whether they exist as a required option or not in the base definition.
+        if (_.has(self.definition.options, label)) {
+            allOptions = allOptions.concat(_.keys(self.definition.options[label]));
+        }
+
+        if (!_.isEmpty(self.definition.options)) {
+            _.forEach(allOptions, function(option) {
+                var taskSpecificOptions = self.definition.options[label];
+                if (_.has(taskSpecificOptions, option)) {
```

Comment:
Though it's not responsible for this PR as it has existed in old code and you just move the code, but I think it's good chance to look at following problem:

This code means only if the user input `option` is in the task definition's options list, then it can be assigned to  the task. But sometimes the task may invent some options when build the workflow, the famous example is the task `When`: https://github.com/RackHD/on-taskgraph/blob/master/lib/graphs/redfish-discovery-graph.js#L13.

This code will just prevent the invented option from passing to the final task.